qiskit/circuit/quantumcircuit.py

```diff
@@ -706,8 +706,15 @@ def compose(self, other, qubits=None, clbits=None, front=False, inplace=False):
         else:
             dest._data += mapped_instrs
 
-        for instr, _, _ in mapped_instrs:
-            dest._update_parameter_table(instr)
+        if front:
+            # rebuild parameter table
+            dest._parameter_table.clear()
```

Comment:
Are parameter tables usually small enough that rebuilding won't be too expensive?

Reply:
The tables themselves should be reasonably small, maybe a few hundred elements at max right now. But you've got a point, it would be better to avoid iterating over all gates again (including the unparameterized ones) so I changed the logic to make it more efficient and only work on the parameter tables.

Reply:
Actually it seems we have to rebuild it since instructions are copied and we cannot use the parameter table from the input `other`. Maybe @kdk has another idea to make it without rebuild 😄

Reply:
We do something similar in `QuantumCircuit.copy`:

https://github.com/Qiskit/qiskit-terra/blob/cb2110957e4b22df8d1bfe42e353fae83af2a14b/qiskit/circuit/quantumcircuit.py#L1686

We make the instruction copies ahead of time so that we can build a map from the old to new instances. Then we can use that map to rebuild the parameter table without having to scan and isinstance all the parameters in the circuit. Maybe a similar approach could be useful here.